File: 两数相除.py
class Solution:
    # 暴力解法，dividend一直减去divisor，次数不断+1
    def divide(self, dividend, divisor):
        """
        :type dividend: int
        :type divisor: int
        :rtype: int
        """
        signal = 1
        if dividend < 0:
            signal *= -1
            dividend *= -1
        if divisor < 0:
            signal *= -1
            divisor *= -1
        count = 0
        while dividend>=divisor:
            dividend -= divisor
            count += 1
        return signal * count

    # 进过优化的方法，相减的次数为1,2,4,8...
    # 例如10/3，第一次10-3，第二次10-6
    # 当然，如果想更加快，可以使用位运算
    def divide_1(self, dividend, divisor):
        if dividend == -pow(2,31) and divisor == -1:
            return pow(2, 31) - 1
        sign = 1
        if (dividend>0 and divisor<0) or (dividend<0 and divisor>0):
            sign = -1
        a, b = abs(dividend), abs(divisor)
        num = 0
        while a >= b:
            sum = b
            count = 1
            while (sum + sum) <= a:
                sum += sum
                count += count
            a = a - sum
            num = num + count
        return num if sign > 0 else -num


a = Solution()
print(a.divide_1(pow(2,30),3))

File: 两数相除.cpp
#include <iostream>
#include <string>
#include <map>
#include <vector>
#include <set>
#include <algorithm>
#include <queue>
#include <unordered_set>
#include <time.h>
#include "linearlist.cpp"

using namespace std;

class Solution {
public:
	int divide(int dividend, int divisor) 
	{
		bool flag = (dividend > 0) ^ (divisor > 0) ? true : false;//判断正负
		//转换为负数，防止越界
		dividend = (dividend > 0) ? -dividend : dividend;
		divisor = (divisor > 0) ? -divisor : divisor;
		int res = 0;//表示输出结果

		while (dividend <= divisor)
		{
			int temp_divisor = divisor, temp_result = -1;//temp_divisor表示除数，temp_result表示移动位数
			while (dividend <= (temp_divisor << 1))
			{
				if (temp_divisor <= (INT_MIN >> 1))
					break;//防越界
				temp_result <<= 1;
				temp_divisor <<= 1;
			}
			dividend -= temp_divisor;
			res += temp_result;
		}
		if (!flag)//负数
		{
			if (res <= INT_MIN)
				return INT_MAX;
			res = -res;
		}
		return res;
	}
};


int main()
{
	Solution a;
	
	cout << a.divide(10,3) << endl;


	system("pause");
}

